pkg/reconciler/pipelinerun/tracing.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerun
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	// Name of the tracer
+	Tracer = "PipelineRunReconciler"
+	// Name of the Annotation for storing SpanContext
+	SpanContextAnnotation = "tekton.dev/pipelinerunSpanContext"
+	// Name of the Annotation used for propogating SpanContext to TaskRun
+	TaskRunSpanContextAnnotation = "tekton.dev/taskrunSpanContext"
+)
+
+// initialize tracing by creating the root span and injecting the spanContext
+// spanContext is propogated through annotations in the CR
+func initTracing(ctx context.Context, tracerProvider trace.TracerProvider, pr *v1beta1.PipelineRun) context.Context {
+	logger := logging.FromContext(ctx)
+	carrier := make(map[string]string)
+
+	pro := otel.GetTextMapPropagator()
+	an := pr.Annotations
+	if an == nil {
+		an = make(map[string]string)
+	}
+
+	// if spanContext is not present in the CR, create new parent trace and
+	// add the marshalled spancontext to the annotations
+	if _, e := an[SpanContextAnnotation]; !e {
+		ctxWithTrace, span := tracerProvider.Tracer(Tracer).Start(ctx, "PipelineRun:Reconciler")
+		defer span.End()
+		span.SetAttributes(attribute.String("pipeline", pr.Name), attribute.String("namespace", pr.Namespace))
+
+		if !span.SpanContext().HasTraceID() {
+			logger.Debug("tracerProvider doesn't provide a traceId, tracing is disabled")
+			return ctx
+		}
+
+		marshalled, err := getMarshalledSpanContext(ctxWithTrace)
+		if err != nil {
+			logger.Error("Unable to marshal spancontext, err: %s", err)
+			return ctx
+		}
+		logger.Info("adding spancontext", marshalled)
+		an[SpanContextAnnotation] = marshalled
+		pr.Annotations = an
+		span.AddEvent("updating TaskRun CR with SpanContext annotations")
+	}
```

Comment:
NIT: I wonder if we could just return the context here, instead of un-marshalling what we just marshalled?

Reply:
ya makes sense